helm-charts/infisical/templates/backend-deployment.yaml

```diff
@@ -34,43 +35,58 @@ spec:
           periodSeconds: 10
         ports: 
         - containerPort: 4000
-        {{- if .Values.backend.kubeSecretRef }}
         envFrom:
         - secretRef:
-            name: {{ .Values.backend.kubeSecretRef }}
-        {{- end }}
-        env:
-        - name: MONGO_URL
-          value: {{ include "infisical.mongodb.connectionString" . | quote }}
-        {{- if .Values.backendEnvironmentVariables }}
-        {{- range $key, $value := .Values.backendEnvironmentVariables }}
-        {{- if $value | quote | eq "MUST_REPLACE" }}
-          {{ fail "Environment variables are not set. Please set all environment variables to continue." }}
-        {{ end }}
-        - name: {{ $key }}
-          value: {{ quote $value }}
-        {{- end }}
-        {{- end }}
+            name: {{ $backend.kubeSecretRef | default (include "infisical.backend.fullname" .) }}
 
 ---
+
 apiVersion: v1
 kind: Service 
 metadata:
   name: {{ include "infisical.backend.fullname" . }}
   labels:
     {{- include "infisical.backend.labels" . | nindent 4 }}
-  {{- with .Values.backend.service.annotations }}
+  {{- with $backend.service.annotations }}
   annotations:
     {{- toYaml . | nindent 4 }}
   {{- end }}
 spec:
-  type: {{ .Values.backend.service.type }}
+  type: {{ $backend.service.type }}
   selector:
     {{- include "infisical.backend.matchLabels" . | nindent 8 }}
   ports:
     - protocol: TCP
       port: 4000
       targetPort: 4000 # container port 
-      {{- if eq .Values.backend.service.type "NodePort" }}
-      nodePort: {{ .Values.backend.service.nodePort }}
+      {{- if eq $backend.service.type "NodePort" }}
+      nodePort: {{ $backend.service.nodePort }}
       {{- end }}
+
+---
+
+{{ if not $backend.kubeSecretRef }}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ include "infisical.backend.fullname" . }}
+  annotations:
+    "helm.sh/resource-policy": "keep"
+type: Opaque
+stringData:
+  {{- $requiredVars := dict "ENCRYPTION_KEY" (randAlphaNum 32 | lower)
+                           "JWT_SIGNUP_SECRET" (randAlphaNum 32 | lower)
+                           "JWT_REFRESH_SECRET" (randAlphaNum 32 | lower)
+                           "JWT_AUTH_SECRET" (randAlphaNum 32 | lower)
+                           "JWT_SERVICE_SECRET" (randAlphaNum 32 | lower)
+                           "JWT_MFA_SECRET" (randAlphaNum 32 | lower) }}
+  {{- $secretObj := (lookup "v1" "Secret" .Release.Namespace (include "infisical.backend.fullname" .)) | default dict }}
+  {{- $secretData := (get $secretObj "data") | default dict }}
+  {{ range $key, $value := .Values.backendEnvironmentVariables }}
+    {{- $default := get $requiredVars $key -}}
+    {{- $current := get $secretData $key | b64dec -}}
+    {{- $v := $value | default ($current | default $default) -}}
+    {{ $key }}: {{ $v | quote }}
+  {{ end -}}
+  MONGO_URL: {{ include "infisical.mongodb.connectionString" . | quote }}
```

Comment:
I assume if the user  placed `MONGO_URL` env under `.Values.backendEnvironmentVariables` then it will be ignored? since  `MONGO_URL: {{ include "infisical.mongodb.connectionString" . | quote }` comes last so it will override it? I wonder if devs expect to just put the mongo url also under `.Values.backendEnvironmentVariables`

Reply:
I'm noticing that when a variable is first initialized with a value then the value is set to "", the K8 secret still has old value and doesn't override to empty value

Reply:
I also expected the deployment to reload when i change values of environment variables in values.yaml file and run helm upgrade but this was not the case. The updates to the secrets don't seem to be fed to the containers

Reply:
Other than some updating cases, the deployment worked with autogenerated vars!! 💥💥

Reply:
> That would be easier to normalize the way we check for existing secrets and user provided values! I could add the MONGO_URL in the required variables dict, auto-generated through the _helpers.tpl and compared against what the user provides. It'll work like every other auto-generated var, same order preferences!
> 
> And I could still add an extra block for backward compatibility, checking the externalMongoDBConnectionString value. to avoid any breaking changes. WDY

Yes! And good catch, let's add the check for externalMongoDBConnectionString to make sure no breaking changes

Reply:
> Did you expect the value to be nulled ? It's not a "bug" it's a "fool-protection feature" 😆

hmm, you are right according to the code this is the expected behavior. I'm just wondering if there would be a reason to set any of the default values to "" but this is most likely a no

Reply:
> That's odd.. it should, I'll double check

sounds good, let me know what you find